Mathlib/Data/Fin/Tuple/Curry.lean

```diff
@@ -17,55 +18,147 @@ n-ary generalizations of the binary `curry` and `uncurry`.
 
 * `Function.OfArity.uncurry`: convert an `n`-ary function to a function from `Fin n → α`.
 * `Function.OfArity.curry`: convert a function from `Fin n → α` to an `n`-ary function.
+* `Function.OfHArity.uncurry`: convert an `p`-ary heterogeneous function to a
+  function from `(i : Fin n) → p i`.
+* `Function.OfHArity.curry`: convert a function from `(i : Fin n) → p i` to a
+  `p`-ary heterogeneous function.
 
 -/
 
-universe u
+universe u v w w'
 
-variable {α β : Type u}
+namespace Function
 
-namespace Function.OfArity
+/-- Currying for 3-ary functions. -/
+@[inline] def curry₃ {α : Type u} {β : Type v} {γ : Type w} {δ : Type w'} :
+    (α × β × γ → δ) → α → β → γ → δ := fun f a b c => f (a, b, c)
+/-- Uncurrying for 3-ary functions. -/
+@[inline] def uncurry₃ {α : Type u} {β : Type v} {γ : Type w} {δ : Type w'} :
+    (α → β → γ → δ) → (α × β × γ) → δ := fun f x => f x.1 x.2.1 x.2.2
```

Comment:
I think these should be reviewed as a separate PR.

Reply:
Sorry, I misunderstood what you were suggesting I take out from the previous PR. I'll strip the 3-currying stuff